srv/cmd/main.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package main implements a simple HTTP/JSON REST example.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/abcxyz/abc-updater/srv/pkg"
+
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/renderer"
+	"github.com/abcxyz/pkg/serving"
+)
+
+const defaultPort = "8080"
+
+var port = flag.String("port", defaultPort, "Specifies server port to listen on.")
+
+// TODO: figure out how to make modules so this doesn't get re-defined multiple places
+type SendMetricRequest struct {
+	// The ID of the application to check.
+	AppID string `json:"appId"`
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	AppVersion string `json:"appVersion"`
+
+	// TODO: this is a bit different from design doc, is it ok?
+	Metrics map[string]int64 `json:"metrics"`
+
+	// InstallID. Expected to be a hex 8-4-4-4-12 formatted v4 UUID.
+	InstallID string `json:"installId"`
+}
+
+type AppMetrics struct {
+	AppID   string
+	Allowed map[string]interface{}
+}
+
+// MetricAllowed is a helper for looking up a particular metric for an app.
+func (m *AppMetrics) MetricAllowed(metric string) bool {
+	if m != nil && m.Allowed != nil {
+		_, ok := m.Allowed[metric]
+		return ok
+	}
+	return false
+}
+
+// GetAllowedMetrics returns a struct containing metrics for a given AppID.
+// An error is returned if that AppID is not defined in the backend for metrics.
+func GetAllowedMetrics(appID string) (*AppMetrics, error) {
+	// TODO: implement me
+	if appID != "implementLater" {
+		return nil, fmt.Errorf("no metric definition found for app %s", appID)
+	}
+	return &AppMetrics{Allowed: map[string]interface{}{"todo-implement": struct{}{}}}, nil
+}
+
+func handleMetric(h *renderer.Renderer) http.Handler {
```

Comment:
We tend to keep main.go clean. I'd suggest moving everything other than `main` and `realMain` to `pkg`.

Reply:
Good point. To avoid git conflict madness (I have a bunch of stacked branches right now) I'll refactor in a new branch at the top of the stack.

Reply:
Branch: pdewilde/refactor-metrics-handler